libraries/iio/iio.c

```diff
@@ -742,32 +722,201 @@ ssize_t iio_step(struct iio_desc *desc)
+{
+	struct iio_channel	*ch;
+	struct iio_attribute	*attr;
+	char			ch_id[50];
+	int32_t			i;
+	int32_t			j;
+	int32_t			k;
+	int32_t			n;
+
+	if (buff_size == -1)
+		n = 0;
+	else
+		n = buff_size;
+
+	if (buff == NULL)
+		/* Set dummy value for buff. It is used only for counting */
+		buff = ch_id;
+
+	i = 0;
+	i += snprintf(buff, max(n - i, 0),
+		      "<device id=\"device%d\" name=\"%s\">", id, name);
+
+	/* Write channels */
+	if (device->channels)
+		for (j = 0; device->channels[j]; j++) {
+			ch = device->channels[j];
+			sprintf(ch_id, "%s%d", get_channel_id(ch->ch_type), j);
+			i += snprintf(buff + i, max(n - i, 0),
+				      "<channel id=\"%s\" name=\"%s\""
+				      " type=\"%s\" >",
+				      ch_id, ch->name,
+				      ch->ch_out ? "output" : "input");
+
+			/* Write channel attributes */
+			if (ch->attributes)
+				for (k = 0; ch->attributes[k]; k++) {
+					attr = ch->attributes[k];
+					i += snprintf(buff + i, max(n - i, 0),
+						      "<attribute name=\"%s\""
+						      " filename=\"%s_%s_%s_%s\" />",
+						      attr->name,
+						      ch->ch_out ? "out" : "in",
+						      ch_id, ch->name,
+						      attr->name);
+				}
+
+			i += snprintf(buff + i, max(n - i, 0), "</channel>");
+		}
+
+	/* Write device attributes */
+	if (device->attributes)
+		for (j = 0; device->attributes[j]; j++)
+			i += snprintf(buff + i, max(n - i, 0),
+				      "<attribute name=\"%s\" />",
+				      device->attributes[j]->name);
+
+	/* Write debug attributes */
+	if (device->debug_attributes)
+		for (j = 0; device->debug_attributes[j]; j++)
+			i += snprintf(buff + i, max(n - i, 0),
+				      "<debug-attribute name=\"%s\" />",
+				      device->debug_attributes[j]->name);
+
+	/* Write buffer attributes */
+	if (device->buffer_attributes)
+		for (j = 0; device->buffer_attributes[j]; j++)
+			i += snprintf(buff + i, max(n - i, 0),
+				      "<buffer-attribute name=\"%s\" />",
+				      device->buffer_attributes[j]->name);
+
+	i += snprintf(buff + i, max(n - i, 0), "</device>");
+
+	return i;
+}
 
 /**
  * @brief Register interface.
- * @param iio_interface - Structure containing physical device instance and
- * device descriptor.
+ * @param desc - iio descriptor
+ * @param dev_descriptor - Device descriptor
+ * @param name - Name to identify the registered device
+ * @param dev_instance - Opened instance of the device
  * @return SUCCESS in case of success or negative value otherwise.
  */
-ssize_t iio_register(struct iio_desc *desc, struct iio_interface *iio_interface)
+ssize_t iio_register(struct iio_desc *desc, struct iio_device *dev_descriptor,
+		     char *name, void *dev_instance)
 {
-	return desc->interfaces_list->push(desc->interfaces_list, iio_interface);
+	struct iio_interface	*iio_interface;
+	int32_t ret;
+	int32_t	n;
+	int32_t	new_size;
+	char	*aux;
+
+	iio_interface = (struct iio_interface *)calloc(1,
+			sizeof(*iio_interface));
+	if (!iio_interface)
+		return -ENOMEM;
+	iio_interface->dev_descriptor = dev_instance;
```

Comment:
the `iio_interface->dev_descriptor` points to the `dev_instance` or `dev_descriptor`?

because on line 844 you set it to `dev_instance`, and immediately after you set it to `dev_descriptor` (line 846)

Reply:
fixed with:
`iio_interface->dev_instance = dev_instance;`